### PR TITLE
[Relay][Runtime] Add support for virtual machine Objects

### PR DESCRIPTION
**Pull request by jroesch**
This PR separates the functionality for passing  `runtime::Object`s around from the VM, VM memory manager and VM compiler.

See #2810 and #2889.

### CODE REVIEW

File: src/runtime/object.cc [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/*!
+ *  Copyright (c) 2017 by Contributors
```
Comment (wweic) [inline src/runtime/object.cc]:
```suggestion
 *  Copyright (c) 2019 by Contributors
```

File: python/tvm/_ffi/function.py [hunk 2]
```diff
@@ -30,19 +30,35 @@
...
     # pylint: disable=wrong-import-position
-    from ._ctypes.function import _set_class_function, _set_class_module
+    from ._ctypes.function import _set_class_function, _set_class_module, _set_class_object
+    from ._ctypes.function import ObjectBase as _ObjectBase
     from ._ctypes.function import FunctionBase as _FunctionBase
     from ._ctypes.function import convert_to_tvm_func
 
+class Object(_ObjectBase):
+    def to_value(self):
+        # fix me
```
Comment (weberlo) [inline python/tvm/_ffi/function.py]: what needs to be fixed?

File: src/runtime/object.cc [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+ * \brief A managed object in the TVM runtime.
+ */
+
+#include <tvm/logging.h>
+#include <tvm/runtime/object.h>
+#include <iostream>
+
+namespace tvm {
+namespace runtime {
+
```
Comment (tqchen) [inline src/runtime/object.cc]: move to src/runtime/vm/object.cc

**Comment by tqchen** [comment]
Some minor nits. After some thoughts, I personally do not like the fact that Node and Object system are separate.  Eventually, we want to merge them into a single node system, but still have a clear separation of compiler and runtime node objects.

For now, we could even reuse the NodeBase class for object(which allows the effective use of make_node). This may need to happen in a few months and I don't want to block the PR, but it might be helpful to keep that in mind when polishing this PR, e.g., the naming(XXXCell-> XXXNode?).

**Comment by tqchen** [comment]
Thanks @jroesch @wweic this is now merged

**Pull request merged**